### PR TITLE
New ItemSelector Widget (SingleItemSelector, MultiItemSelector) - libyui part

### PR DESCRIPTION
**Pull request by shundhammer**
## Trello

https://trello.com/c/7qI1AAvN/1300-5-libyui-1-of-n-selection-widget

## Bugzilla

- Bug: System roles are not properly rendered in graphical installation

  https://bugzilla.suse.com/show_bug.cgi?id=1084674

- Bug: selection box auto scrolls incorrectly

  https://bugzilla.suse.com/show_bug.cgi?id=1083207

- Bug: UX regression in system role selection (mainly ncurses)

  https://bugzilla.suse.com/show_bug.cgi?id=1086187

## Mockup Screenshot

time-warp-02

## Description

This is a new widget for a scrollable 1-of-n or n-of-m selection with not only a simple text for each item, but a heading text and a (possibly multi-line, even rich text) description.

This is the libyui part of it, providing the base class `YItemSelector` and factory methods in `YWidgetFactory`.

## Other Changes

- Code cleanup: Cleaned up 10 years of cruft in the UI (incompatible indentation etc.).

- Reverted insane change of `std::` everything for common data types even in the `.cc` files. There is absolutely no need to clutter even the C++ implementation files with that; it just makes the code less readable. Now again using plain `string` and `endl` in the `.cc` files again instead of `std::string` and `std::endl`. This commit from 2012 should never have been accepted.

  While it might make sense to some extent for `.h` files to not have `using std::something`, it is completely insane to demand that in the `.cc` files as well. Even for `.h` files this is questionable: Nobody with sense invents a new `string` class without  prefix or whatever because this will automatically clash with `std::string` in a gazillion places.

- Moved out some of the helper functions from `YDialog.cc` to a new file `YDialogHelpers.cc`. This was completely misplaced there to begin with. `YDialog` is one of the most important classes in this library; it is counterproductive to clutter it with unrelated stuff like a convenience dialog that can show help texts or release notes. This needs to go to a separate class soonish. `YDialog` is much too important to have it cluttered with stuff like that.

## Related PRs

- libyui-qt: https://github.com/libyui/libyui-qt/pull/115

### CODE REVIEW

File: src/YItemSelector.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+     *
+     * Reimplemented from YWidget.
+     **/
+    virtual const YPropertySet & propertySet();
+
+    /**
+     * The name of the widget property that will return user input.
+     * Inherited from YWidget.
+     **/
+    const char * userInputProperty() { return YUIProperty_Value; }
```
Comment (aschnell) [inline src/YItemSelector.h]: Several functions here should be marked as ```override```.

File: src/YDialogHelpers.cc [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+/-*/
+
+
+#include <sys/types.h>
+#include <dirent.h>
+#include <fstream>
+
+#include <list>
+#include <algorithm>
```
Comment (aschnell) [inline src/YDialogHelpers.cc]: I cannot see the need for the includes above.

File: src/YDescribedItem.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+     * Destructor.
+     **/
+    virtual ~YDescribedItem() {}
+
+    /**
+     * Return this item's description text. This is the (typically longer)
+     * subtext that the user sees in a dialog, so this will usually be a
+     * translated text.
+     **/
+    std::string description() const { return _description; }
```
Comment (aschnell) [inline src/YDescribedItem.h]: Right now this could return a const reference.

File: src/YDialogHelpers.cc [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    {
+	map<string,string> relnotes = YUI::application()->releaseNotes();
+
+	if ( relnotes.size() == 0)
+	{
+	    return false;
+	}
+
+	vector<string> keys;
+	for ( map<string,string>::iterator it = relnotes.begin(); it != relnotes.end(); ++it )
```
Comment (aschnell) [inline src/YDialogHelpers.cc]: This could be a ```const_iterator``` (already in the original location).

File: src/YItemSelector.h (hunk 1 above)
Comment (shundhammer) [reply src/YItemSelector.h]: In theory yes; but then we should do that consistently all over this library. We don't use that anywhere else so far, so introducing it here would be inconsistent.

Some day we should bring the benefits of C++11 to this library, but then everywhere in a consistent way.

File: src/YDescribedItem.h (hunk 3 above)
Comment (shundhammer) [reply src/YDescribedItem.h]: This is also a matter of consistency; the rest of the library does not do that.

File: src/YDialogHelpers.cc (hunk 4 above)
Comment (shundhammer) [reply src/YDialogHelpers.cc]: That code was just moved verbatim to a new file.

**Comment by shundhammer** [comment]
As discussed in the daily meeting, bumping the version to 3.8.0. I had to force-push for that.

File: src/YDialogHelpers.cc (hunk 2 above)
Comment (shundhammer) [reply src/YDialogHelpers.cc]: Fixed.

File: src/YDialogHelpers.cc (hunk 4 above)
Comment (shundhammer) [reply src/YDialogHelpers.cc]: Fixed.

**Pull request merged**